src/sentry/sentry_apps/logic.py

```diff
@@ -206,30 +210,62 @@ def _update_events(self) -> None:
-            return
-
-        create_or_update_service_hooks_for_installation(
-            installation=installation,
-            webhook_url=self.sentry_app.webhook_url,
-            events=self.sentry_app.events,
+        organization_context = organization_service.get_organization_by_id(
+            id=self.sentry_app.owner_id
         )
+        assert organization_context is not None, "Organization cannot be None for ff"
+        if features.has(
+            "organizations:service-hooks-outbox",
+            organization_context.organization,
+        ):
+            installations = SentryAppInstallation.objects.filter(sentry_app_id=self.sentry_app.id)
+
+            org_ids_and_region_names = OrganizationMapping.objects.filter(
+                organization_id__in=installations.values_list("organization_id", flat=True)
+            ).values_list("organization_id", "region_name")
+
+            installation_org_id_to_region_name = {
+                org_id: region_name for org_id, region_name in org_ids_and_region_names
+            }
+
+            with outbox_context(
+                transaction.atomic(router.db_for_write(ControlOutbox)), flush=False
+            ):
+                for installation in installations:
+                    ControlOutbox(
+                        shard_scope=OutboxScope.APP_SCOPE,
+                        shard_identifier=self.sentry_app.id,
+                        object_identifier=installation.id,
+                        category=OutboxCategory.SERVICE_HOOK_UPDATE,
+                        region_name=installation_org_id_to_region_name[
+                            installation.organization_id
+                        ],
+                    ).save()
+        else:
+            if self.sentry_app.is_published:
+                # if it's a published integration, we need to do many updates so we have to do it in a task so we don't time out
+                # the client won't know it succeeds but there's not much we can do about that unfortunately
+                create_or_update_service_hooks_for_sentry_app.apply_async(
+                    kwargs={
+                        "sentry_app_id": self.sentry_app.id,
+                        "webhook_url": self.sentry_app.webhook_url,
+                        "events": self.sentry_app.events,
+                    }
+                )
+                return
+
+            # for unpublished integrations that aren't installed yet, we may not have an installation
+            # if we don't, then won't have any service hooks
+            try:
+                installation = SentryAppInstallation.objects.get(sentry_app_id=self.sentry_app.id)
+            except SentryAppInstallation.DoesNotExist:
+                return
+
+            create_or_update_service_hooks_for_installation(
+                installation=installation,
+                webhook_url=self.sentry_app.webhook_url,
+                events=self.sentry_app.events,
```

Comment:
Breaking this out into its own helper method would make this a bit cleaner. Likewise, breaking out the new logic into its own method would also make this more readable imo.